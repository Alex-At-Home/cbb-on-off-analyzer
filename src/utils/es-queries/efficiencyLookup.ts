import _ from 'lodash';
import { HighMajorConfs } from "../public-data/ConferenceInfo"

/** Returns the full set of information from the efficiency lookup primarily from KenPom (with permission) */
export const efficiencyLookup = function(
    year: String,
    nameLookup: Record<string, Record<string, string>> //(from public-data/ncaaToKenomLookup)
) {
    //POST kenpom_all/_search
    return {
        "from": "0",
        "_source": {
           "includes": [
              "ncaa_name",
              "team_season.year",
              "conf",
              "stats.adj_off",
              "stats.adj_def",
              "stats.adj_margin",
              "stats.adj_tempo",
              "stats.off._3p_pct.value",
              "total_poss",
              "ncaa_seed"
           ],
           "excludes": []
        },
        "sort": [
           {
              "stats.adj_margin.rank": {
                 "order": "asc"
              }
           }
        ],
        "size": "1000",
        "query": {
           "bool": {
              "filter": [],
              "must_not": [],
              "should": [],
              "must": [
                 {
                    "query_string": {
                       "query": `team_season.year:${year}` //(eg 2022 for 21/22)
                    }
                 }
              ]
           }
        },
        "script_fields": {
           "ncaa_name": {
              "script": {
                 "lang": "painless",
                 "source": "// Use doc and params, return the field value\ndef team = doc[\"team_season.team.keyword\"].value;\ndef year = doc[\"team_season.year\"].value.toString();\ndef paramKey = \"kp_to_pbp\";\ndef pbp_team = params[paramKey][team];\nif (null != pbp_team) {\n   return pbp_team.NCAA_name;\n} else {\n   return team;\n}",
                 "params": {
                    "kp_to_pbp": nameLookup,
                 }
              }
           },
           "total_poss": {
              "script": {
                 "lang": "painless",
                 "source": "// Use doc and params, return the field value\ndef total_poss = 0;\nfor (p in doc[\"games.pace\"]) {\n   total_poss += p;\n}\nreturn total_poss;",
                 "params": {}
              }
           }
        }
     }
} 

export const formatEfficiencyLookupResponse = function(response: any) {
    return _.chain(response?.hits?.hits || []).map(team => {
        const teamSource = team._source || {};
        const teamFields = team.fields || {};
        const conf = teamSource?.conf || "";
        return [teamFields?.ncaa_name?.[0] || "unknown", {
            "team_season.year": teamSource?.team_season?.year || "",
            "conf": conf,
            "stats.adj_off.rank": teamSource?.stats?.adj_off?.rank || 0,
            "stats.adj_off.value": teamSource?.stats?.adj_off?.value || 0,
            "stats.adj_def.rank": teamSource?.stats?.adj_def?.rank || 0,
            "stats.adj_def.value": teamSource?.stats?.adj_def?.value || 0,
            "stats.adj_margin.rank": teamSource?.stats?.adj_margin?.rank || 0,
            "stats.adj_margin.value": teamSource?.stats?.adj_margin?.value || 0,
            "stats.adj_tempo.rank": teamSource?.stats?.adj_tempo?.rank || 0,
            "stats.adj_tempo.value": teamSource?.stats?.adj_tempo?.value || 0,
            "stats.off._3p_pct.value": teamSource?.stats?.off?._3p_pct?.value || 0,
            "total_poss": teamFields?.total_poss?.[0] || 0,
            "ncaa_seed": teamSource?.ncaa_seed || undefined,
            "is_high_major": HighMajorConfs.has(conf) ? 1 : 0 
        }];
    }).fromPairs().value();
}