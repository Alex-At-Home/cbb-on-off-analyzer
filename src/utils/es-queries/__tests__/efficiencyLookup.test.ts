
import { efficiencyLookup, formatEfficiencyLookupResponse } from "../efficiencyLookup";

describe("efficiencyLookup", () => {

  expect.addSnapshotSerializer({
    test: (val: any) => true,
    print: (val: any) => JSON.stringify(val, null, 3)
  });

  test("efficiencyLookup", () => {

    expect(efficiencyLookup("2021", { test: { field: "value" } })).toMatchSnapshot();
  });
  test("formatEfficiencyLookupResponse", () => {
    const input = {
      "took": 59,
      "timed_out": false,
      "_shards": {
        "total": 5,
        "successful": 5,
        "skipped": 0,
        "failed": 0
      },
      "hits": {
        "total": {
          "value": 348,
          "relation": "eq"
        },
        "max_score": null,
        "hits": [{
          "_index": "kenpom_all",
          "_type": "doc",
          "_id": "Gonzaga 2021",
          "_score": null,
          "_source": {
            "ncaa_seed": 1,
            "stats": {
              "adj_off": {
                "rank": 1,
                "value": 126.4
              },
              "adj_tempo": {
                "rank": 7,
                "value": 73.8
              },
              "adj_def": {
                "rank": 11,
                "value": 89.9
              },
              "off": {
                "_3p_pct": {
                  "value": 36.8
                }
              },
              "adj_margin": {
                "rank": 1,
                "value": 36.5
              }
            },
            "team_season": {
              "year": 2021
            },
            "conf": "West Coast Conference"
          },
          "fields": {
            "total_poss": [2387],
            "ncaa_name": ["Gonzaga"]
          },
          "sort": [1]
        }, {
          "_index": "kenpom_all",
          "_type": "doc",
          "_id": "Baylor 2021",
          "_score": null,
          "_source": {
            //removed seed for test purposes
            "stats": {
              "adj_off": {
                "rank": 2,
                "value": 125
              },
              "adj_tempo": {
                "rank": 213,
                "value": 67.4
              },
              "adj_margin": {
                "rank": 2,
                "value": 33.900000000000006
              },
              "adj_def": {
                "rank": 22,
                "value": 91.1
              },
              "off": {
                "_3p_pct": {
                  "value": 41.3
                }
              }
            },
            "team_season": {
              "year": 2021
            },
            "conf": "Big 12 Conference"
          },
          "fields": {
            "total_poss": [2086],
            "ncaa_name": ["Baylor"]
          },
          "sort": [2]
        }, {
          "_index": "kenpom_all",
          "_type": "doc",
          "_id": "Michigan 2021",
          "_score": null,
          "_source": {
            "ncaa_seed": 1,
            "stats": {
              "adj_off": {
                "rank": 9,
                "value": 117.6
              },
              "adj_tempo": {
                "rank": 256,
                "value": 66.6
              },
              "adj_margin": {
                "rank": 3,
                "value": 29.69999999999999
              },
              "adj_def": {
                "rank": 4,
                "value": 87.9
              },
              "off": {
                "_3p_pct": {
                  "value": 38.1
                }
              }
            },
            "team_season": {
              "year": 2021
            },
            "conf": "Big Ten Conference"
          },
          "fields": {
            "total_poss": [1902],
            "ncaa_name": ["Michigan"]
          },
          "sort": [3]
        }]
      }
    }
    expect(formatEfficiencyLookupResponse(input)).toMatchSnapshot();
  });

});
