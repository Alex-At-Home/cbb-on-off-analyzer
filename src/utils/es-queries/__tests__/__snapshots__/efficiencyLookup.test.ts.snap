// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`efficiencyLookup efficiencyLookup 1`] = `
{
   "from": "0",
   "_source": {
      "includes": [
         "ncaa_name",
         "team_season.year",
         "conf",
         "stats.adj_off",
         "stats.adj_def",
         "stats.adj_margin",
         "stats.adj_tempo",
         "stats.off._3p_pct.value",
         "total_poss",
         "ncaa_seed"
      ],
      "excludes": []
   },
   "sort": [
      {
         "stats.adj_margin.rank": {
            "order": "asc"
         }
      }
   ],
   "size": "1000",
   "query": {
      "bool": {
         "filter": [],
         "must_not": [],
         "should": [],
         "must": [
            {
               "query_string": {
                  "query": "team_season.year:2021"
               }
            }
         ]
      }
   },
   "script_fields": {
      "ncaa_name": {
         "script": {
            "lang": "painless",
            "source": "// Use doc and params, return the field value\\ndef team = doc[\\"team_season.team.keyword\\"].value;\\ndef year = doc[\\"team_season.year\\"].value.toString();\\ndef paramKey = \\"kp_to_pbp\\";\\ndef pbp_team = params[paramKey][team];\\nif (null != pbp_team) {\\n   return pbp_team.NCAA_name;\\n} else {\\n   return team;\\n}",
            "params": {
               "kp_to_pbp": {
                  "test": {
                     "field": "value"
                  }
               }
            }
         }
      },
      "total_poss": {
         "script": {
            "lang": "painless",
            "source": "// Use doc and params, return the field value\\ndef total_poss = 0;\\nfor (p in doc[\\"games.pace\\"]) {\\n   total_poss += p;\\n}\\nreturn total_poss;",
            "params": {}
         }
      }
   }
}
`;

exports[`efficiencyLookup formatEfficiencyLookupResponse 1`] = `
{
   "Gonzaga": {
      "team_season.year": 2021,
      "conf": "West Coast Conference",
      "stats.adj_off.rank": 1,
      "stats.adj_off.value": 126.4,
      "stats.adj_def.rank": 11,
      "stats.adj_def.value": 89.9,
      "stats.adj_margin.rank": 1,
      "stats.adj_margin.value": 36.5,
      "stats.adj_tempo.rank": 7,
      "stats.adj_tempo.value": 73.8,
      "stats.off._3p_pct.value": 36.8,
      "total_poss": 2387,
      "ncaa_seed": 1,
      "is_high_major": 0
   },
   "Baylor": {
      "team_season.year": 2021,
      "conf": "Big 12 Conference",
      "stats.adj_off.rank": 2,
      "stats.adj_off.value": 125,
      "stats.adj_def.rank": 22,
      "stats.adj_def.value": 91.1,
      "stats.adj_margin.rank": 2,
      "stats.adj_margin.value": 33.900000000000006,
      "stats.adj_tempo.rank": 213,
      "stats.adj_tempo.value": 67.4,
      "stats.off._3p_pct.value": 41.3,
      "total_poss": 2086,
      "is_high_major": 1
   },
   "Michigan": {
      "team_season.year": 2021,
      "conf": "Big Ten Conference",
      "stats.adj_off.rank": 9,
      "stats.adj_off.value": 117.6,
      "stats.adj_def.rank": 4,
      "stats.adj_def.value": 87.9,
      "stats.adj_margin.rank": 3,
      "stats.adj_margin.value": 29.69999999999999,
      "stats.adj_tempo.rank": 256,
      "stats.adj_tempo.value": 66.6,
      "stats.off._3p_pct.value": 38.1,
      "total_poss": 1902,
      "ncaa_seed": 1,
      "is_high_major": 1
   }
}
`;
